packages/ui-lib/src/Alerts/atom/Tooltip.tsx

```diff
@@ -0,0 +1,321 @@
+              top: ${currentTop}px;
+              left: ${currentLeft - ARROW_SIZE - ARROW_MARGIN + (width / 2)}px;
+              transform: translateY(-100%);
+              ${Arrow} {
+                transform: rotate(180deg);
+                margin-left: ${ARROW_MARGIN}px;
+              }
+            `;
+
+    case 'top-center':
+      return `
+              flex-direction: column-reverse;
+              align-items: center;
+              top: ${currentTop}px;
+              left: ${currentLeft + (width / 2)}px;
+              transform: translate(-50%, -100%);
+              ${Arrow} {
+                transform: rotate(180deg);
+              }
+            `;
+
+    case 'top-start':
+      return `
+              flex-direction: column-reverse;
+              align-items: end;
+              top: ${currentTop}px;
+              left: ${currentLeft + ARROW_SIZE + ARROW_MARGIN + (width / 2)}px;
+              transform: translate(-100%, -100%);
+              ${Arrow} {
+                transform: rotate(180deg);
+                margin-right: ${ARROW_MARGIN}px;
+              }
+            `;
+
+    // default case is 'bottom-center'
+    default:
+      return `
+            flex-direction: column;
+            top: ${currentTop + height}px;
+            left: ${currentLeft + (width / 2)}px;
+            transform: translateX(-50%);
+            align-items: center;
+          `;
+  }
+};
+
+const getDynamicPosition = (coords: DOMRect, width?: number, height?: number): ITooltipPosition => {
+
+  const isOverflowingRight = coords.left + (width || 0) + 30 > window.innerWidth;
+  const isOverflowingBottom = coords.bottom + (height || 0) > window.innerHeight;
+  const isOverflowingLeft = coords.left - (width || 0) < 0;
+  const isOverflowingTop = coords.top - (height || 0) < 0;
+
+  if (isOverflowingTop && isOverflowingRight && isOverflowingBottom) {
+    return 'left-center';
+  }
+
+  if (isOverflowingTop && isOverflowingLeft && isOverflowingBottom) {
+    return 'right-center';
+  }
+
+  if (isOverflowingTop && isOverflowingLeft) {
+    return 'bottom-end';
+  }
+
+  if (isOverflowingTop && isOverflowingRight) {
+    return 'bottom-start';
+  }
+
+  if (isOverflowingBottom && isOverflowingLeft) {
+    return 'top-end';
+  }
+
+  if (isOverflowingBottom && isOverflowingRight) {
+    return 'top-start';
+  }
+
+  if (isOverflowingLeft) {
+    return 'right-center';
+  }
+
+  if (isOverflowingRight) {
+    return 'left-center';
+  }
+
+  if (isOverflowingBottom) {
+    return 'top-center';
+  }
+
+  return 'bottom-center';
+};
+
+type ITooltip = {
+  message: string
+  tooltipFor: string
+  icon?: string
+  type?: ITooltipType
+  tooltipPosition?: ITooltipPosition
+  maxWidth?: string
+}
+
+export type ITooltipPosition = 'top-start' | 'top-center' | 'top-end' | 'bottom-start' | 'bottom-center' | 'bottom-end' | 'left-start' | 'left-center' | 'left-end' | 'right-start' | 'right-center' | 'right-end';
```

Comment:
Reviewing it again now we have all the options, I feel like definitions here are a little confusing. It could be a bug but I think it might just be the naming isn't clear. I've tried to select the option in Storybook that I'm looking for and often got it wrong.

For example, if I choose `LeftStart` I am expecting it to be on the left (which it is) and I am expecting the anchor point to be at the top of the tooltip (as the start), so the start of the tooltip aligns with the start of the trigger element. At the moment the vertical position is the opposite of what I'd expect.

Thinking about it, most of the uses will be centered. My thinking is we split the props to take advantage of this. The first required prop would be: `top`, `right`, `bottom` and `left`. These would all default to their center variants. Then I think a secondary prop would allow us to select a position offset `start`, `center` or `end`. When the `start` is used, the anchor point would be at the `left` or `top` with the tooltip aligned to that. `end` would refer to `bottom` or `right` being the anchor point.

That's just my proposal for the props, internally the naming doesn't need refactoring unless you feel it advantagous.

Drop me a message if you want to chat through things!

Reply:
Hello, 

I think we have different visions of the naming and interactions of the tooltip for the developer, let's chat to align expectations.

Reply:
> Reviewing it again now we have all the options, I feel like definitions here are a little confusing. It could be a bug but I think it might just be the naming isn't clear. I've tried to select the option in Storybook that I'm looking for and often got it wrong.
> 
> For example, if I choose `LeftStart` I am expecting it to be on the left (which it is) and I am expecting the anchor point to be at the top of the tooltip (as the start), so the start of the tooltip aligns with the start of the trigger element. At the moment the vertical position is the opposite of what I'd expect.
> 
> Thinking about it, most of the uses will be centered. My thinking is we split the props to take advantage of this. The first required prop would be: `top`, `right`, `bottom` and `left`. These would all default to their center variants. Then I think a secondary prop would allow us to select a position offset `start`, `center` or `end`. When the `start` is used, the anchor point would be at the `left` or `top` with the tooltip aligned to that. `end` would refer to `bottom` or `right` being the anchor point.
> 
> That's just my proposal for the props, internally the naming doesn't need refactoring unless you feel it advantagous.
> 
> Drop me a message if you want to chat through things!

I have updated the positions as TopLeft manner like we agree in chat, please do a sanity check review.